libfwupdplugin/fu-struct.c

```diff
@@ -0,0 +1,1465 @@
+/*
+ * Copyright (C) 2023 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#define G_LOG_DOMAIN "FuStruct"
+
+#include "config.h"
+
+#include <gio/gio.h>
+#include <string.h>
+
+#include "fu-byte-array.h"
+#include "fu-common.h"
+#include "fu-mem.h"
+#include "fu-string.h"
+#include "fu-struct.h"
+
+/**
+ * FuStruct:
+ *
+ * A smart structure that supports endian conversion, arrays, GUIDs, strings, default and constant
+ * data of variable size.
+ *
+ * In most cases the smart structure will be defined in the #GObject `_init()` function:
+ *
+ *    fu_struct_register(self,
+ *                       "StructureName {"
+ *                       "    member_id: u16le,"
+ *                       "}");
+ *
+ * ..and then the structure can be used in parsing functions like this:
+ *
+ *    FuStruct *st = fu_struct_lookup(self, "StructureName");
+ *    if (!fu_struct_unpack_full(st, buf, bufsz, offset, error))
+ *        return FALSE;
+ *    value = fu_struct_get_u16(st, "member_id");
+ *
+ * ..and binary writing functions like this:
+ *
+ *    FuStruct *st = fu_struct_lookup(self, "StructureName");
+ *    fu_struct_set_u16(st, "member_id", value);
+ *    fu_struct_pack_into(st, buf);
+ *
+ * NOTE: is is very important to only read and write the correct member ID type and non-constant
+ * values. For instance, it is invalid to read a `u16le` type into a `guint32` value using
+ * `fu_struct_get_u32()`.
```

Comment:
It's going to be harder to catch bugs because of this I think.

Reply:
The other thing we could do is put the blobs of struct into a `fu-fmap-firmware.struct` file and then process them at build time to have `guint16 fu_fmap_header_get_nareas(st);`, `fu_fmap_header_parse(st, buf, bufsz, error)`, `GByteArray fu_fmap_header_pack(st)` in generated fu-fmap-header.c and fu-fmap-header..h files. That's dangerously close to just re-inventing a subset of protobuf but without any of the chaos. If you think that makes sense say and I can prototype something.

Reply:
Why not just use protobuf then?

Reply:
I spent the morning trying to use protobuf. I've opened https://github.com/fwupd/fwupd/pull/5639 with a similar idea that might work better for us.